CIP-bytestring-map-builtin/README.md

```diff
@@ -0,0 +1,152 @@
+
+Another attempt to mitigate the performance restrictions imposed by the list implementation was to specialise the Plutus Core implementation [into an association list with `ByteString` keys][4]. Unfortunately, this approach also failed.
+
+We concluded that embedding maps into Plutus Core cannot be done without incurring high performance costs.
+
+### Map as a builtin type
+
+The only remaining solution is to consider introducing a builtin map type.
+
+Unfortunately, the builtin language has a significant limitation: it does not allow for higher-order builtin functions. As of the time of writing, builtin functions may not receive other functions as arguments, but there is ongoing work to lift this restriction. However, due to complex design considerations, repeated applications of function arguments will incur high performance costs. Since maps are recursive data structures, these high performance costs will be accumulated and such an approach will still not be feasible.
+
+Why are higher-order builtin functions necessary? Map keys must be processed (or simply, just compared) in order to be able to build and operate over well-defined maps. For arbitrary key types, that would only be possible by passing the required processing or comparison functions as arguments to the builtin functions.
+
+The only way to implement a builtin map with a (builtin) polymorphic key type would be to modify the builtins language itself. A compiler would have to only allow the construction of maps with key types which are known a priori to posess some property (such as a total ordering). The Plutus Core evaluator would have to check for this as well. The implementation itself would not be too complicated, but we must first consider how this would change the formal semantics of the builtins language.
+
+Given the above, the only immediate solution is to implement a builtin map type which is specialised to `ByteString` keys but remains polymorphic in the value type. We believe that this solution should be good enough for most practical use-cases, and maybe even surpass a fully polymorphic map type in performance.
+
+
+## Specification
+
+We propose the addition of the following builtin type: `ByteStringMap` of kind `Type -> Type`.
+- It represents a map associating keys of type `ByteString` with values of any builtin type.
+- The new builtin type should be implemented using a tree data structure allowing for logarithmic-time map operations.
+- We consider the costing size of a `ByteStringMap` to be equal to the size of its corresponding representation as an association list.
```

Comment:
I think you could simply say "roughly equal" instead of "equal" and that'll be fine, this is all to be figured out during the costing phase anyway, it's fine not to have the exact answer now.

Reply:
> @kwxm what should we say here?

@ana: sorry, I missed this.  We actually have to caculate the size of the map object before we carry out any operations on it, and that might be expensive if we have to traverse a tree to  findi ts size.  One way to alleviate this might be to have the maps carry their sizes about with them and make sure that the update operations update the size as well (and during deserialisation we'd have to check that maps embedded in scripts don't lie about their size).  We might even need two measures of size: one telling you the number of elements of the map (for converting to a list, for example), and one telling you the depth of the map (for insertion etc.). We could do this by using newtype wrappers to choose which size measure should be applied.  Maybe we could just store the size of the object and take the log instead of measuring the depth, since these should be proportional if the tree's always kept balanced.  This is all probably too detailed for a CIP though, so saying that the size should be something like the number of entries should be fine: we can work out the details later.